conbench/api/results.py

```diff
@@ -157,96 +137,108 @@ def get(self) -> f.Response:
+        filters = []
+        joins = []
+
+        if run_id_arg := f.request.args.get("run_id"):
+            filters.append(BenchmarkResult.run_id == run_id_arg)
+        else:
+            filters.append(BenchmarkResult.timestamp >= "2023-06-03")
 
         if name_arg := f.request.args.get("name"):
-            benchmark_results = BenchmarkResult.search(
-                filters=[Case.name == name_arg] + days_filters,
-                order_by=BenchmarkResult.timestamp.desc(),
-                joins=[Case],
-                limit=limit,
-            )
+            filters.append(Case.name == name_arg)
+            joins.append(Case)
 
-        elif batch_id_arg := f.request.args.get("batch_id"):
-            batch_ids = batch_id_arg.split(",")
-            benchmark_results = BenchmarkResult.search(
-                filters=[BenchmarkResult.batch_id.in_(batch_ids)]
-            )
+        if batch_id_arg := f.request.args.get("batch_id"):
+            filters.append(BenchmarkResult.batch_id == batch_id_arg)
+
+        if run_reason_arg := f.request.args.get("run_reason"):
+            filters.append(BenchmarkResult.run_reason == run_reason_arg)
 
-        elif run_id_arg := f.request.args.get("run_id"):
-            run_ids = run_id_arg.split(",")
-            if len(run_ids) > 5:
-                return resp400(
-                    "it is currently not allowed to set more than five "
-                    "run_id values; consider making separate requests (see "
-                    "issue 978)"
-                )
-
-            benchmark_results = current_session.scalars(
-                select(BenchmarkResult).where(BenchmarkResult.run_id.in_(run_ids))
-            ).all()
-
-        elif run_reason_arg := f.request.args.get("run_reason"):
-            if not limit:
-                limit = 30000
-            benchmark_results = BenchmarkResult.search(
-                filters=[BenchmarkResult.run_reason == run_reason_arg] + days_filters,
-                order_by=BenchmarkResult.timestamp.desc(),
-                limit=limit,
+        cursor_arg = f.request.args.get("cursor")
+        if cursor_arg and cursor_arg != "null":
+            filters.append(BenchmarkResult.id < cursor_arg)
+
+        page_size = f.request.args.get("page_size", 100)
+        try:
+            page_size = int(page_size)
+            assert 1 <= page_size <= 1000
+        except Exception:
+            self.abort_400_bad_request(
+                "page_size must be a positive integer no greater than 1000"
             )
 
+        query = select(BenchmarkResult)
+        for join in joins:
+            query = query.join(join)
+        query = (
+            query.filter(*filters).order_by(BenchmarkResult.id.desc()).limit(page_size)
+        )
+        benchmark_results = current_session.scalars(query).all()
+
+        if len(benchmark_results) == page_size:
+            next_page_cursor = benchmark_results[-1].id
```

Comment:
There's a corner case here where the last page has exactly `page_size` results. I guess the call with that cursor would just return an empty list and maybe that's fine? But it's a wasted call.

Reply:
Yep! I thought about this and decided it's worth the tradeoff. Either:
1. the client makes one redundant request whenever the last page has `page_size` results (probably on average about 1 in `page_size` times? so 1% of the time, or 0.1% for clients maxing out the size)
2. the server makes another DB query every single time just to make sure the next page is empty

(1) sounds like a lot less server load. I'll add a comment about this though.